editoast/src/views/timetable/paced_train.rs

```diff
@@ -722,31 +732,68 @@ async fn occupancy_blocks(
         })
         .await?;
 
-    let first_occurrences = paced_trains
-        .clone()
-        .into_iter()
-        .map(|p| p.into_train_schedule())
-        .collect_vec();
+    let simulation_contexts: Vec<SimulationContext> =
+        paced_trains
+            .iter()
+            .flat_map(|paced_train| {
+                std::iter::once(SimulationContext {
+                    paced_train_id: paced_train.id,
+                    exception_key: None,
+                    train_schedule: paced_train.clone().into_train_schedule(),
+                })
+                .chain(paced_train.exceptions.iter().map(|exception| {
+                    SimulationContext {
+                        paced_train_id: paced_train.id,
+                        exception_key: Some(exception.key.clone()),
+                        train_schedule: paced_train.apply_exception(exception),
+                    }
+                }))
+            })
+            .collect();
 
     let occupancy_blocks_result = compute_occupancy_blocks(
         conn,
         core_client,
         valkey_client,
         path,
         infra,
-        first_occurrences,
+        simulation_contexts
+            .iter()
+            .map(|c| c.train_schedule.clone())
+            .collect(),
```

Comment:
That's expensive... It seems `compute_occupancy_blocks()` could take a `&[models::TrainSchedule]` instead.